telemetry/derive/src/lib.rs

```diff
@@ -0,0 +1,213 @@
+use quote::{quote, ToTokens};
+use syn::{
+    parse::Parse, parse_macro_input, punctuated::Punctuated, token::Comma, FnArg, ItemFn, LitStr,
+    Path, Type,
+};
+
+fn type_has_metrics_field(ty: &Type) -> bool {
+    match ty {
+        // This may seem fragile, but it isn't. We use the same
+        // convention everywhere in the code base, and if you follow
+        // `CONTRIBUTING.md` you'll likely have `use
+        // iroha_data_model::WorldStateView` somewhere.
+        Type::Path(pth) => {
+            let Path { segments, .. } = pth.path.clone();
+            let type_name = &segments[0].ident;
+            *type_name == "WorldStateView"
+        }
+        _ => false,
+    }
+}
+
+/// The identifier of the first argument that has a type which has
+/// metrics.
+///
+/// # Errors
+/// If no argument is of type `WorldStateView`.
+fn arg_metrics(input: &Punctuated<FnArg, Comma>) -> Result<syn::Ident, &Punctuated<FnArg, Comma>> {
+    input
+        .iter()
+        .find(|arg| match arg {
+            FnArg::Typed(typ) => match &*typ.ty {
+                syn::Type::Reference(typ) => type_has_metrics_field(&typ.elem),
+                _ => false,
+            },
+            _ => false,
+        })
+        .and_then(|arg| match arg {
+            FnArg::Typed(typ) => match *typ.pat.clone() {
+                syn::Pat::Ident(ident) => Some(ident.ident),
+                _ => None,
+            },
+            _ => None,
+        })
+        .ok_or(input)
+}
+
+struct MetricSpecs(Vec<MetricSpec>); // `HashSet` — idiomatic; slow
+
+impl Parse for MetricSpecs {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let vars = Punctuated::<MetricSpec, Comma>::parse_terminated(input)?;
+        Ok(Self(vars.into_iter().collect()))
+    }
+}
+
+struct MetricSpec {
+    timing: bool,
+    metric_name: LitStr,
+}
+
+impl Parse for MetricSpec {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let timing = <syn::Token![+]>::parse(input).is_ok();
+        let metric_name_lit = syn::Lit::parse(input)?;
+        let metric_name = match metric_name_lit {
+            syn::Lit::Str(lit_str) => lit_str,
+            _ => {
+                return Err(syn::Error::new(
+                    proc_macro2::Span::call_site(),
+                    "Must be a string literal. Format `[+]\"name_of_metric\"`.",
+                ))
+            }
+        };
+        Ok(Self {
+            metric_name,
+            timing,
+        })
+    }
+}
+
+impl ToTokens for MetricSpec {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        self.metric_name.to_tokens(tokens)
+    }
+}
+
+/// Macro for instrumenting an `isi`'s `impl execute` to track a given
+/// metric.  To specify a metric, put it as an attribute parameter
+/// inside quotes. If you also want to track the execution time of the
+/// `isi`, you should prefix the quoted metric with the `+` symbol.
```

Comment:
Rather confusing syntax to my mind. Why not have 2 macros:
`metric` and `metric_with_time`?

Reply:
Example: 
```
#[metrics(+"some_metric", "expensive_metric_without_timing", +"cheap_metric", "another_expensive_metric")]
```

If it can be all done in one macro, there's less code to maintain, and the compile times are not that adversely affected. I originally wanted to prefix the string with the identifier `timed"some_metric"` but that both looked ugly. + looks very much like a t, so i thought it'd be a good middle-ground.

Reply:
Ok, I agree that using + here is ok